File: haskell/ss5/sortsum2.hs
import Data.List
import Data.Array

sortsums xs ys = map fst (sortsubs xs (map negate ys))
sortsubs xs ys = sortBy (cmp (mkArray xs ys)) (subs xs ys)
subs xs ys = [(x - y, (i,j))|(x,i) <- zip xs [1..], (y,j) <- zip ys [1..]]
cmp a (x,(i,j)) (y,(k,l)) = compare (a ! (1,i,k)) (a ! (2,j,l))
mkArray xs ys = array b (zip (table xs ys) [1..])
                where b = ((1,1,1),(2,p,p))
                      p = max (length xs) (length ys)
table xs ys   = map snd (map (tag 1) xxs `merge` map (tag 2) yys)
                where xxs = sortsubs' xs
                      yys = sortsubs' ys
tag i (x,(j,k)) = (x,(i,j,k))

sortsubs' [] = []
sortsubs' [w] = [(w - w,(1,1))]
sortsubs' ws = foldr1 merge [xxs, map (incr m) xys,
                             map (incl m) yxs, map (incb m) yys]
  where xxs = sortsubs' xs
        xys = sortBy (cmp (mkArray xs ys)) (subs xs ys)
        yxs = map switch (reverse xys)
        yys = sortsubs' ys
        (xs,ys) = splitAt m ws
        m = length ws `div` 2

incl m (x,(i,j)) = (x,(m+i,j))
incr m (x,(i,j)) = (x,(i,m+j))
incb m (x,(i,j)) = (x,(m+i,m+j))
switch (x,(i,j)) = (negate x,(j,i))

merge :: (Ord a) => [a] -> [a] -> [a]
merge [] ys = ys
merge xs [] = xs
merge (x:xs) (y:ys) | x < y     = x : merge xs (y:ys)
                    | otherwise = y : merge (x:xs) ys

File: src/main/scala/examples/ss5/Sortsum2.scala
package examples.ss5

object Sortsum2 {
  
  type Label = (Int,(Int,Int))
  type TaggedLabel = (Int,(Int,Int,Int))
  type LabeledArray = Map[(Int,Int,Int), Int]
    
  def sortsums(xs: List[Int], ys: List[Int]) = sortsubs(xs, ys.map(_ * -1)).map(_._1)

  def sortsubs(xs: List[Int], ys: List[Int]):List[Label] = subs(xs,ys).sortWith(cmp(mkArray(xs,ys)))
    
  def subs(xs: List[Int], ys: List[Int]): List[Label] = {
    for {
      (i, x) <- Stream.from(1).zip(xs).toList
      (j, y) <- Stream.from(1).zip(ys).toList
    } yield (x - y, (i,j))
  }
  
  def cmp(a:LabeledArray)(l1:Label, l2:Label): Boolean = (l1,l2) match {
    case ((x,(i,j)),(y,(k,l))) => a((1,i,k)) - a((2,j,l)) < 0
  }

  def mkArray(xs:List[Int], ys:List[Int]): LabeledArray = {
    println((xs,ys))
    val m = table(xs,ys).zip(Stream.from(1)).toMap
    println(m)
    m
  }
  
  def table(xs: List[Int], ys: List[Int]):List[(Int,Int,Int)] = {
    val xxs:List[TaggedLabel] = sortsubs2(xs).map(tag(1,_))
    val yys:List[TaggedLabel] = sortsubs2(ys).map(tag(2,_))
    merge(xxs, yys).map(_._2)
  } 

  def merge[A: Ordering](list1: List[A], list2: List[A]): List[A] = {
    (list1,list2) match {
      case (Nil, ys) => ys
      case (xs, Nil) => xs
      case (x::xs, y::ys) => 
        if (implicitly[Ordering[A]].lt(x,y)) x :: merge(xs, y::ys)
        else y :: merge(x::xs, ys)
    }
  } 
  
  def tag(i:Int, x: Label): TaggedLabel = x match {case (x,(j,k)) => (x,(i,j,k))}
    //(x._1, (i,x._2._1, x._2._1))

  def switch:PartialFunction[Label,Label] = {case (x:Int,(i,j)) => (x * -1, (j,i))}
  
  def sortsubs2(ws: List[Int]):List[Label] = {
    lazy val m = ws.length / 2
    lazy val (xs,ys) = ws.splitAt(m)
    lazy val yys = sortsubs2(ys)
    lazy val xys = subs(xs,ys).sortWith(cmp(mkArray(xs,ys)))
    lazy val yxs = xys.reverse.map(switch)
    lazy val xxs = sortsubs2(xs)
    ws match {
      case Nil => Nil
      case w::Nil => List((w-w,(1,1)))
      case ws => val list = List(xys.map(incr(m, _)), yxs.map(incl(m, _)), yys.map(incb(m, _)))
                 list.foldRight(xxs)(merge)
    }
  }
  
  def incl(m:Int, a: Label): Label = {
    val (x, (i,j)) = a
    (x,(m+i,j))
  }
  
  def incr(m:Int, a: Label): Label = {
    val (x, (i,j)) = a
    (x,(i,m+j))
  }
  
  def incb(m:Int, a: Label): Label = {
    val (x, (i,j)) = a
    (x,(m+i,m+j))
  }
  
  def switch(a: Label): Label = {
    val (x, (i,j)) = a
    (x * -1,(i,j))
  }
}